app/scripts/metamask-controller.js

```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
     }
   }
 
+  // eth_decryptMessage methods
+
+  /**
+  * Called when a dapp uses the eth_decryptMessage method.
+  *
+  * @param {Object} msgParams - The params of the message to sign & return to the Dapp.
+  * @param {Function} cb - The callback function called with the signature.
+  * Passed back to the requesting Dapp.
+  */
+  async newRequestDecryptMessage (msgParams, req) {
+    const promise = this.decryptMessageManager.addUnapprovedMessageAsync(msgParams, req)
+    this.sendUpdate()
+    this.opts.showUnconfirmedMessage()
+    return promise
+  }
+
+  /**
+  * Only decypt message and don't touch transaction state
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessageInline (msgParams) {
+    log.info('MetaMaskController - decryptMessageInline')
+    const msgId = msgParams.metamaskId
+    const stripped = ethUtil.stripHexPrefix(msgParams.data)
+    const buff = Buffer.from(stripped, 'hex')
+    msgParams.data = JSON.parse(buff.toString('utf8'))
+    // decrypt the message inline
+    return this.keyringController.decryptMessage(msgParams).then((rawData) => {
+      const msg = this.decryptMessageManager.getMsg(msgId)
+      msg.rawData = rawData
+      this.decryptMessageManager._updateMsg(msg)
+      return this.getState()
+    })
+  }
+
+  /**
+  * Signifies a user's approval to decrypt a message in queue.
+  * Triggers decrypt, and the callback function from newUnsignedDecryptMessage.
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessage (msgParams) {
+    log.info('MetaMaskController - decryptMessage')
+    const msgId = msgParams.metamaskId
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.decryptMessageManager.approveMessage(msgParams)
+        .then((cleanMsgParams) => {
+          const stripped = ethUtil.stripHexPrefix(cleanMsgParams.data)
+          const buff = Buffer.from(stripped, 'hex')
+          cleanMsgParams.data = JSON.parse(buff.toString('utf8'))
+          // decrypt the message
+          return this.keyringController.decryptMessage(cleanMsgParams)
+        })
+        .then((rawMess) => {
+        // tells the listener that the message has been decrypted
+        // and can be returned to the dapp
+          this.decryptMessageManager.setMsgStatusDecrypted(msgId, rawMess)
+          return this.getState()
+        })
+    } catch (error) {
+      log.info('MetaMaskController - eth_decryptMessage failed.', error)
+      this.decryptMessageManager.errorMessage(msgId, error)
```

Comment:
This should probably re-throw the error, so that the UI knows something wrong happened.